action/protocol/staking/handler_stake_migrate.go

```diff
@@ -0,0 +1,208 @@
+package staking
+
+import (
+	"context"
+	"math/big"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/pkg/errors"
+	"go.uber.org/zap"
+
+	"github.com/iotexproject/iotex-address/address"
+	"github.com/iotexproject/iotex-proto/golang/iotextypes"
+
+	"github.com/iotexproject/iotex-core/action"
+	"github.com/iotexproject/iotex-core/action/protocol"
+	accountutil "github.com/iotexproject/iotex-core/action/protocol/account/util"
+	"github.com/iotexproject/iotex-core/action/protocol/execution"
+	"github.com/iotexproject/iotex-core/pkg/log"
+	"github.com/iotexproject/iotex-core/pkg/util/byteutil"
+	"github.com/iotexproject/iotex-core/state"
+)
+
+func (p *Protocol) handleStakeMigrate(ctx context.Context, act *action.MigrateStake, csm CandidateStateManager) ([]*action.Log, []*action.TransactionLog, uint64, uint64, error) {
+	actLogs := make([]*action.Log, 0)
+	transferLogs := make([]*action.TransactionLog, 0)
+	insGas, err := act.IntrinsicGas()
+	if err != nil {
+		return nil, nil, 0, 0, err
+	}
+	gasConsumed := insGas
+	gasToBeDeducted := insGas
+	bucket, rErr := p.fetchBucket(csm, act.BucketIndex())
+	if rErr != nil {
+		return nil, nil, gasConsumed, gasToBeDeducted, rErr
+	}
+	staker, rerr := fetchCaller(ctx, csm, big.NewInt(0))
+	if rerr != nil {
+		return nil, nil, gasConsumed, gasToBeDeducted, errors.Wrap(rerr, "failed to fetch caller")
+	}
+	candidate := csm.GetByIdentifier(bucket.Candidate)
+	if candidate == nil {
+		return nil, nil, gasConsumed, gasToBeDeducted, errCandNotExist
+	}
+	duration := uint64(bucket.StakedDuration / p.helperCtx.BlockInterval(protocol.MustGetBlockCtx(ctx).BlockHeight))
+	exec, err := p.constructExecution(candidate.GetIdentifier(), bucket.StakedAmount, duration, act.Nonce(), act.GasLimit(), act.GasPrice())
+	if err != nil {
+		return nil, nil, gasConsumed, gasToBeDeducted, errors.Wrap(err, "failed to construct execution")
+	}
+	// validate bucket index
+	if err := p.validateStakeMigrate(ctx, bucket, csm); err != nil {
+		return nil, nil, gasConsumed, gasToBeDeducted, err
+	}
+
+	// snapshot for sm in case of failure of hybrid protocol handling
+	si := csm.SM().Snapshot()
+	revertSM := func() {
+		if revertErr := csm.SM().Revert(si); revertErr != nil {
+			log.L().Panic("failed to revert state", zap.Error(revertErr))
+		}
+	}
+	// force-withdraw native bucket
+	actLog, tLog, err := p.withdrawBucket(ctx, staker, bucket, candidate, csm)
+	if err != nil {
+		return nil, nil, gasConsumed, gasToBeDeducted, err
```

Comment:
no need to revertSM() here?
if the last error to put action caller into SM failed, it won't cause change to SM?